pkg/updatestatus/nodeinformer.go

```diff
@@ -0,0 +1,390 @@
+	mcdState := node.Annotations[mco.MachineConfigDaemonStateAnnotationKey]
+	return isUpdating && mcdState == mco.MachineConfigDaemonStateDone
+}
+
+func determineConditions(pool *machineconfigv1.MachineConfigPool, node *corev1.Node, isUpdating, isUpdated, isUnavailable, isDegraded bool, lns *mco.LayeredNodeState, now metav1.Time) ([]metav1.Condition, string, *metav1.Duration) {
+	var estimate *metav1.Duration
+	var message string
+
+	updating := metav1.Condition{
+		Type:               string(NodeStatusInsightUpdating),
+		Status:             metav1.ConditionUnknown,
+		Reason:             string(NodeCannotDetermine),
+		LastTransitionTime: now,
+	}
+	available := metav1.Condition{
+		Type:               string(NodeStatusInsightAvailable),
+		Status:             metav1.ConditionTrue,
+		LastTransitionTime: now,
+	}
+	degraded := metav1.Condition{
+		Type:               string(NodeStatusInsightDegraded),
+		Status:             metav1.ConditionFalse,
+		LastTransitionTime: now,
+	}
+
+	if isUpdating && isNodeDraining(node, isUpdating) {
+		estimate = toPointer(10 * time.Minute)
+		updating.Status = metav1.ConditionTrue
+		updating.Reason = string(NodeDraining)
+	} else if isUpdating {
+		state := node.Annotations[mco.MachineConfigDaemonStateAnnotationKey]
+		switch state {
+		case mco.MachineConfigDaemonStateRebooting:
+			estimate = toPointer(10 * time.Minute)
+			updating.Status = metav1.ConditionTrue
+			updating.Reason = string(NodeRebooting)
+		case mco.MachineConfigDaemonStateDone:
+			estimate = toPointer(time.Duration(0))
+			updating.Status = metav1.ConditionFalse
+			updating.Reason = string(NodeCompleted)
+		default:
+			estimate = toPointer(10 * time.Minute)
+			updating.Status = metav1.ConditionTrue
+			updating.Reason = string(NodeUpdating)
+		}
+
+	} else if isUpdated {
+		estimate = toPointer(time.Duration(0))
+		updating.Status = metav1.ConditionFalse
+		updating.Reason = string(NodeCompleted)
+	} else if pool.Spec.Paused {
+		estimate = toPointer(time.Duration(0))
+		updating.Status = metav1.ConditionFalse
+		updating.Reason = string(NodePaused)
+	} else {
+		updating.Status = metav1.ConditionFalse
+		updating.Reason = string(NodeUpdatePending)
+	}
+	message = updating.Message
+
+	if isUnavailable && !isUpdating {
+		estimate = nil
+		if isUpdated {
+			estimate = toPointer(time.Duration(0))
+		}
+		available.Status = metav1.ConditionFalse
+		available.Reason = lns.GetUnavailableReason()
+		available.Message = lns.GetUnavailableMessage()
+		available.LastTransitionTime = metav1.Time{Time: lns.GetUnavailableSince()}
+		message = available.Message
+	}
+
+	if isDegraded {
+		estimate = nil
+		if isUpdated {
+			estimate = toPointer(time.Duration(0))
+		}
+		degraded.Status = metav1.ConditionTrue
+		degraded.Reason = node.Annotations[mco.MachineConfigDaemonReasonAnnotationKey]
+		degraded.Message = node.Annotations[mco.MachineConfigDaemonReasonAnnotationKey]
+		message = degraded.Message
```

Comment:
I'm not sure if overwriting the message is the right thing to do, ideally the message somehow captures the important information from all three sources (but we need to take length constraints into account)